rfcs/20240718-synthetic-benchmark/README.md

```diff
@@ -0,0 +1,279 @@
+syntax:
+
+```Bash
+collectdnn --<primitive> --validate --report=<metric> --samples=<num> --<filter_cfg1>=<param1> --<filter_cfg2>=<param2> ...
+```
+
+As a psuedo-code summary for how this will be implemented,
+
+
+```C++
+Generator G(primtitive_kind, filter_cfg, enable_auto_diff);
+Results R;
+std::pair<Problem, Primitive> P;
+parallel_for(Problem p : G) {
+    P.emplace_back({p, create_primitive(p)});
+}
+
+for(auto p : P) {
+    if (benchmark) r.emplace_back(benchmark(p, metric, ctx));
+    if(validate) p.validate();
+}
+
+```
+
+As this tool is intended to produce data from multiple oneDNN builds, this tool
+should rely on benchdnn for data generation. To properly support data
+collection, benchdnn will need two new features. The first feature is to enable
+logging and skipping problems based off of a primitive ID. These IDs can then be
+used for the automated diffing feature discussed above.
+
+There are two main options for implementing primtive IDs
+ * Option 1: Serialize relevant primitive data
+ * Option 2: Hash relevant primitive data to generate a UUID
+
+Given that we are looking at generating millions of primitives, and serialized
+primitive data is likely to be in the 10's of kilobytes in size, Option 1
+appears infeasible due to the data size. While using a hash allows for false
+collisions, by choosing a large enough key, we are unlikely to hit an issue in
+practice. To implement this, we would add the following internal interface for
+use in benchdnn:
+
+`extern "C" primitive_d_t DNNL_API dnnl_primitive_get_id(primitive_iface_t *p). `
+
+
+To provide a maintainable and correct implementation, `dnnl_primitive_get_id` is
+expected to provide a has of the primitive JIT generated function binary and any
+performance related function arguments. As getting the hash of runtime kernel
+arguments will need to be invasive to each implementation, a special value
+`hash_id_unsupported = 0` will also exist for unsupported implementations,
+so that developer can focus on implementations under active development.
+
+Finally, we will need to add a new hashing mode-modifier to enable logging of
+primitive ID's, along with a new input `--skip-primitive-ids=<file>`. If we
+implement a hashing mode denoted as `H`, automatic diffing can then be
+implemented as:
+
+``` Bash
+    ./build1/tests/benchdnn/benchdnn --mode=H ... > skip_list
+    ./build2/tests/benchdnn/benchdnn --skip-primitives=skip_list ...
+```
+
+A prototype implementation of the proposed `--mode=H` option for GPU primitives
+is contained on the branch `rjoursle/diffdnn`.
+
+The second modification that will be needed for benchdnn is a new low-fidelity
+batched benchmarking mode. This mode will would only execute each workload once,
+with the expectation that run-to-run variability is being handled by executing
+multiple similar workloads. The benchmarking is expected to be performed as
+follows on GPU, although some tweaks may be required to get performance stability
```

Comment:
The problem with the `--fix-times-per-prb` knob, is that it is not setup to keep the device saturated with work. This could cause issues with benchmarking if the target device enters some kind power saving mode while oneDNN primitives are being created. To avoid that, benchdnn currently performs a warmup phase for each workload. This warmup should be unnecessary in a batched benchmarking mode and as the goal here is to execute a primitive exactly once, the existing warmup phase would be a significant contributor to benchmarking throughput.